File: complexipy/utils/sarif.py
from __future__ import annotations

import json
from importlib.metadata import PackageNotFoundError
from importlib.metadata import version as pkg_version
from typing import List

from complexipy._complexipy import FileComplexity

_RULE_ID = "CC001"
_SCHEMA = (
    "https://raw.githubusercontent.com/oasis-tcs/sarif-spec"
    "/master/Schemata/sarif-schema-2.1.0.json"
)
_INFO_URI = "https://rohaquinlop.github.io/complexipy/"
_HELP_URI = "https://rohaquinlop.github.io/complexipy/understanding-scores/"


def _get_version() -> str:
    try:
        return pkg_version("complexipy")
    except PackageNotFoundError:
        return "unknown"


def _refactor_plan_level(applicability: object) -> str:
    return "note" if "Informational" in str(applicability) else "warning"


def _refactor_plan_result(file: FileComplexity, function, plan) -> dict:
    return {
        "ruleId": plan.rule_id,
        "level": _refactor_plan_level(plan.applicability),
        "message": {"text": f"{plan.title}: {plan.explanation}"},
        "locations": [
            {
                "physicalLocation": {
                    "artifactLocation": {
                        "uri": file.path,
                        "uriBaseId": "%SRCROOT%",
                    },
                    "region": {
                        "startLine": int(plan.line_start),
                        "startColumn": int(plan.column_start),
                        "endLine": int(plan.line_end),
                    },
                },
                "logicalLocations": [
                    {"name": function.name, "kind": "function"}
                ],
            }
        ],
    }


def _refactor_plan_rule_definition(plan) -> dict:
    return {
        "id": plan.rule_id,
        "name": plan.kind,
        "shortDescription": {"text": plan.description},
        "helpUri": plan.doc_url,
        "properties": {"tags": [str(plan.category).lower()]},
    }


def _complexity_result(
    file: FileComplexity, function, max_complexity: int
) -> dict:
    return {
        "ruleId": _RULE_ID,
        "level": "warning",
        "message": {
            "text": (
                f"Function '{function.name}' has a cognitive complexity"
                f" of {function.complexity}, which exceeds the maximum"
                f" allowed complexity of {max_complexity}."
            )
        },
        "locations": [
            {
                "physicalLocation": {
                    "artifactLocation": {
                        "uri": file.path,
                        "uriBaseId": "%SRCROOT%",
                    },
                    "region": {
                        "startLine": int(function.line_start),
                        "endLine": int(function.line_end),
                    },
                },
                "logicalLocations": [
                    {"name": function.name, "kind": "function"}
                ],
            }
        ],
    }


def _complexity_rule_definition() -> dict:
    return {
        "id": _RULE_ID,
        "name": "CognitiveComplexity",
        "shortDescription": {"text": "Cognitive complexity exceeds threshold"},
        "helpUri": _HELP_URI,
        "properties": {"tags": ["maintainability", "readability"]},
    }


def _add_refactor_plan_results(
    file: FileComplexity,
    function,
    results: list,
    refactor_rule_definitions: dict,
) -> None:
    for plan in function.refactor_plans:
        if plan.rule_id not in refactor_rule_definitions:
            refactor_rule_definitions[plan.rule_id] = (
                _refactor_plan_rule_definition(plan)
            )
        results.append(_refactor_plan_result(file, function, plan))


def _collect_results(
    files: List[FileComplexity], max_complexity: int, suggest_refactors: bool
) -> tuple:
    results: list = []
    refactor_rule_definitions: dict = {}

    for file in files:
        for function in file.functions:
            if function.complexity > max_complexity:
                results.append(
                    _complexity_result(file, function, max_complexity)
                )
            if suggest_refactors:
                _add_refactor_plan_results(
                    file, function, results, refactor_rule_definitions
                )

    return results, refactor_rule_definitions


def store_sarif(
    output_path: str,
    files: List[FileComplexity],
    max_complexity: int,
    suggest_refactors: bool = False,
) -> None:
    """Write complexity violations as a SARIF 2.1.0 file.

    Functions whose complexity exceeds *max_complexity* are always emitted as
    SARIF results under the built-in cognitive-complexity rule. When
    *suggest_refactors* is True, each function's ranked refactor plans are
    also emitted as results under their own rule ID (C001, C007, ...), with
    rule definitions built dynamically from the plans actually encountered so
    the catalog can never drift from `RuleMetadata`. The file is written to
    *output_path*.
    """
    results, refactor_rule_definitions = _collect_results(
        files, max_complexity, suggest_refactors
    )

    sarif_doc = {
        "version": "2.1.0",
        "$schema": _SCHEMA,
        "runs": [
            {
                "tool": {
                    "driver": {
                        "name": "complexipy",
                        "version": _get_version(),
                        "informationUri": _INFO_URI,
                        "rules": [
                            _complexity_rule_definition(),
                            *refactor_rule_definitions.values(),
                        ],
                    }
                },
                "results": results,
            }
        ],
    }

    with open(output_path, "w", encoding="utf-8") as f:
        json.dump(sarif_doc, f, indent=2)

File: crates/complexipy-cli/src/utils/sarif.rs
use std::fs;

use serde_json::{Map, Value, json};

use complexipy_core::classes::{
    Applicability, FileComplexity, FunctionComplexity, RefactorPlan, RuleCategory,
};
use complexipy_core::utils::ExportError;

const RULE_ID: &str = "CC001";
const SCHEMA: &str = "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json";
const INFO_URI: &str = "https://rohaquinlop.github.io/complexipy/";
const HELP_URI: &str = "https://rohaquinlop.github.io/complexipy/understanding-scores/";

pub fn store_sarif(
    output_path: &str,
    files: &[FileComplexity],
    max_complexity: u64,
    suggest_refactors: bool,
) -> Result<(), ExportError> {
    let (results, refactor_rule_definitions) =
        collect_results(files, max_complexity, suggest_refactors);

    let mut rules = vec![complexity_rule_definition()];
    rules.extend(refactor_rule_definitions.values().cloned());

    let sarif_doc = json!({
        "version": "2.1.0",
        "$schema": SCHEMA,
        "runs": [{
            "tool": {
                "driver": {
                    "name": "complexipy",
                    "version": env!("CARGO_PKG_VERSION"),
                    "informationUri": INFO_URI,
                    "rules": rules,
                }
            },
            "results": results,
        }],
    });

    let serialized = serde_json::to_string_pretty(&sarif_doc)
        .map_err(|e| ExportError::Serialize(format!("Failed to serialize SARIF: {}", e)))?;
    fs::write(output_path, serialized)
        .map_err(|e| ExportError::Io(format!("Failed to write SARIF to {}: {}", output_path, e)))?;
    Ok(())
}

fn collect_results(
    files: &[FileComplexity],
    max_complexity: u64,
    suggest_refactors: bool,
) -> (Vec<Value>, Map<String, Value>) {
    let mut results = Vec::new();
    let mut refactor_rule_definitions = Map::new();

    for file in files {
        for function in &file.functions {
            if function.complexity > max_complexity {
                results.push(complexity_result(file, function, max_complexity));
            }
            if suggest_refactors {
                add_refactor_plan_results(
                    file,
                    function,
                    &mut results,
                    &mut refactor_rule_definitions,
                );
            }
        }
    }

    (results, refactor_rule_definitions)
}

fn add_refactor_plan_results(
    file: &FileComplexity,
    function: &FunctionComplexity,
    results: &mut Vec<Value>,
    refactor_rule_definitions: &mut Map<String, Value>,
) {
    for plan in &function.refactor_plans {
        if !refactor_rule_definitions.contains_key(&plan.rule_id) {
            refactor_rule_definitions
                .insert(plan.rule_id.clone(), refactor_plan_rule_definition(plan));
        }
        results.push(refactor_plan_result(file, function, plan));
    }
}

fn complexity_result(
    file: &FileComplexity,
    function: &FunctionComplexity,
    max_complexity: u64,
) -> Value {
    json!({
        "ruleId": RULE_ID,
        "level": "warning",
        "message": {
            "text": format!(
                "Function '{}' has a cognitive complexity of {}, which exceeds the maximum allowed complexity of {}.",
                function.name, function.complexity, max_complexity
            )
        },
        "locations": [{
            "physicalLocation": {
                "artifactLocation": {
                    "uri": file.path,
                    "uriBaseId": "%SRCROOT%",
                },
                "region": {
                    "startLine": function.line_start,
                    "endLine": function.line_end,
                },
            },
            "logicalLocations": [
                {"name": function.name, "kind": "function"}
            ],
        }],
    })
}

fn complexity_rule_definition() -> Value {
    json!({
        "id": RULE_ID,
        "name": "CognitiveComplexity",
        "shortDescription": {"text": "Cognitive complexity exceeds threshold"},
        "helpUri": HELP_URI,
        "properties": {"tags": ["maintainability", "readability"]},
    })
}

fn refactor_plan_result(
    file: &FileComplexity,
    function: &FunctionComplexity,
    plan: &RefactorPlan,
) -> Value {
    json!({
        "ruleId": plan.rule_id,
        "level": refactor_plan_level(&plan.applicability),
        "message": {"text": format!("{}: {}", plan.title, plan.explanation)},
        "locations": [{
            "physicalLocation": {
                "artifactLocation": {
                    "uri": file.path,
                    "uriBaseId": "%SRCROOT%",
                },
                "region": {
                    "startLine": plan.line_start,
                    "startColumn": plan.column_start,
                    "endLine": plan.line_end,
                },
            },
            "logicalLocations": [
                {"name": function.name, "kind": "function"}
            ],
        }],
    })
}

fn refactor_plan_rule_definition(plan: &RefactorPlan) -> Value {
    json!({
        "id": plan.rule_id,
        "name": plan.kind,
        "shortDescription": {"text": plan.description},
        "helpUri": plan.doc_url,
        "properties": {"tags": [rule_category_tag(&plan.category)]},
    })
}

fn refactor_plan_level(applicability: &Applicability) -> &'static str {
    match applicability {
        Applicability::Informational => "note",
        _ => "warning",
    }
}

fn rule_category_tag(category: &RuleCategory) -> &'static str {
    match category {
        RuleCategory::Complexity => "complexity",
        RuleCategory::Readability => "readability",
    }
}

#[cfg(test)]
mod tests;
